src/java.base/share/classes/jdk/internal/module/ModuleInfoWriter.java

```diff
@@ -82,86 +81,81 @@ private ModuleInfoWriter() { }
     private static byte[] toModuleInfo(ModuleDescriptor md,
                                        ModuleResolution mres,
                                        ModuleTarget target) {
-        ClassWriter cw = new ClassWriter(0);
-        cw.visit(Opcodes.V10, ACC_MODULE, "module-info", null, null, null);
-
-        int moduleFlags = md.modifiers().stream()
-                .map(MODULE_MODS_TO_FLAGS::get)
-                .reduce(0, (x, y) -> (x | y));
-        String vs = md.rawVersion().orElse(null);
-        ModuleVisitor mv = cw.visitModule(md.name(), moduleFlags, vs);
-
-        // requires
-        for (ModuleDescriptor.Requires r : md.requires()) {
-            int flags = r.modifiers().stream()
-                    .map(REQUIRES_MODS_TO_FLAGS::get)
-                    .reduce(0, (x, y) -> (x | y));
-            vs = r.rawCompiledVersion().orElse(null);
-            mv.visitRequire(r.name(), flags, vs);
-        }
-
-        // exports
-        for (ModuleDescriptor.Exports e : md.exports()) {
-            int flags = e.modifiers().stream()
-                    .map(EXPORTS_MODS_TO_FLAGS::get)
-                    .reduce(0, (x, y) -> (x | y));
-            String[] targets = e.targets().toArray(EMPTY_STRING_ARRAY);
-            mv.visitExport(e.source().replace('.', '/'), flags, targets);
-        }
-
-        // opens
-        for (ModuleDescriptor.Opens opens : md.opens()) {
-            int flags = opens.modifiers().stream()
-                    .map(OPENS_MODS_TO_FLAGS::get)
-                    .reduce(0, (x, y) -> (x | y));
-            String[] targets = opens.targets().toArray(EMPTY_STRING_ARRAY);
-            mv.visitOpen(opens.source().replace('.', '/'), flags, targets);
-        }
-
-        // uses
-        md.uses().stream().map(sn -> sn.replace('.', '/')).forEach(mv::visitUse);
-
-        // provides
-        for (ModuleDescriptor.Provides p : md.provides()) {
-            mv.visitProvide(p.service().replace('.', '/'),
-                            p.providers()
-                                .stream()
-                                .map(pn -> pn.replace('.', '/'))
-                                .toArray(String[]::new));
+        try {
+            return Classfile.buildModule(
+                ModuleAttribute.of(ModuleDesc.of(md.name()), mb -> {
+                    mb.moduleFlags(md.modifiers().stream()
+                            .map(mm -> MODULE_MODS_TO_FLAGS.getOrDefault(mm, 0))
```

Comment:
Should we use `mapToInt` instead of a plain `map`? Same for the other flags below.

Reply:
yes, `mapToInt` seems to be better, thanks.